hera_cal/redcal.py

```diff
@@ -269,6 +337,127 @@ def multiply_by_gains(target, gains, target_type='vis'):
+        to linearize via the Omnical algorithm from Liu et al. 2010.
+
+        Args:
+            data: Dictionary that maps nonlinear product equations, written as valid python-interpetable 
+                strings that include the variables in question, to (complex) numbers or numpy arrarys. 
+                Variables with trailing underscores '_' are interpreted as complex conjugates (e.g. x*y_ 
+                parses as x * y.conj()).
+            sol0: Dictionary mapping all variables (as keyword strings) to their starting guess values.
+                This is the point that is Taylor expanded around, so it must be relatively close to the
+                true chi^2 minimizing solution. In the same format as that produced by 
+                linsolve.LogProductSolver.solve() or linsolve.LinProductSolver.solve().
+            wgts: Dictionary that maps equation strings from data to real weights to apply to each 
+                equation. Weights are treated as 1/sigma^2. All equations in the data must have a weight 
+                if wgts is not the default, {}, which means all 1.0s.
+            gain: The fractional step made toward the new solution each iteration.
+            **kwargs: keyword arguments of constants (python variables in keys of data that 
+                are not to be solved for)
+
+        Returns:
+            None
+        """
+        linsolve.LinProductSolver.__init__(self, data, sol0, wgts=wgts, **kwargs)
+        self.gain = np.float32(gain) # float32 to avoid accidentally promoting data to doubles.
+
+    def _get_ans0(self, sol, keys=None):
+        '''Evaluate the system of equations given input sol. 
+        Specify keys to evaluate only a subset of the equations.'''
+        if keys is None:
+            keys = self.keys
+        _sol = {k+'_':v.conj() for k,v in sol.items() if k.startswith('g')}
+        _sol.update(sol)
+        return {k: eval(k, _sol) for k in keys}
+
+    def solve_iteratively(self, conv_crit=1e-10, maxiter=50, check_every=4, check_after=1, verbose=False):
+        """Repeatedly solves and updates solution until convergence or maxiter is reached. 
+        Returns a meta object containing the number of iterations, chisq, and convergence criterion.
+
+        Args:
+            conv_crit: A convergence criterion (default 1e-10) below which to stop iterating. 
+                Converegence is measured L2-norm of the change in the solution of all the variables
+                divided by the L2-norm of the solution itself.
+            maxiter: An integer maximum number of iterations to perform before quitting. Default 50.
+            check_every: Compute convergence every Nth iteration (saves computation).  Default 4.
+            check_after: Start computing convergence only after N iterations.  Default 1.
+
+        Returns: meta, sol
+            meta: a dictionary with metadata about the solution, including
```

Comment:
I'd clarify that each of these has the shape of the data.

Reply:
Done.